### PR TITLE
Added methods in memory library that gets and sets a transparent hugepage values

### PR DESCRIPTION
**Pull request by santhosh93**
Also write_file_or_log has been added in genio library which writes
to a file and raises exception when the value cannot be written into
the file or if the file does not exists.

Signed-off-by: Santhosh G <santhog4@linux.vnet.ibm.com>

### CODE REVIEW

File: avocado/utils/genio.py [hunk 1]
```diff
@@ -174,3 +174,15 @@ def write_one_line(filename, line):
     :type line: str
     """
     write_file(filename, line.rstrip('\n') + '\n')
+
+
+class GenIOException(Exception):
+    pass
+
+    def write_file_or_log(self, filename, data):
```
Comment (PraveenPenguin) [inline avocado/utils/genio.py]: @santhosh93 i think write_file_or_log can be moved out from class definition

File: avocado/utils/memory.py [hunk 2]
```diff
@@ -265,8 +267,8 @@ def get_buddy_info(chunk_sizes, nodes="all", zones="all"):
 
     if re.findall("[<>=]", chunk_sizes) and buddy_list:
         size_list = range(len(buddy_list[-1][-1].strip().split()))
-        chunk_sizes = [str(_) for _ in size_list if eval("%s %s" % (_,
-                                                                    chunk_sizes))]
+        chunk_sizes = [str(_) for _ in size_list if eval("%s %s" %
+                                                         (_, chunk_sizes))]
```
Comment (PraveenPenguin) [inline avocado/utils/memory.py]: why this change is required ?

File: avocado/utils/memory.py [hunk 3]
```diff
@@ -278,3 +280,34 @@ def get_buddy_info(chunk_sizes, nodes="all", zones="all"):
...
+
+    :param feature : Thp feature to set
+    :type feature : str
+    :param value : Value to be set to feature
+    :type value : str
+    """
+    thp_path = '/sys/kernel/mm/transparent_hugepage/'
+    thp_feature_to_set = os.path.join(thp_path, feature)
+    io = GenIOException()
+    io.write_file_or_log(thp_feature_to_set, value)
```
Comment (PraveenPenguin) [inline avocado/utils/memory.py]: it can be like

`genio.write_file_or_log(thp_feature_to_set, value)`

File: avocado/utils/memory.py (hunk 2 above)
Comment (santhosh93) [inline avocado/utils/memory.py]: actually it throws out a pep8 error like it crosses 80 character in my local system...so moved it to next line...

**Comment by santhosh93** [comment]
@PraveenPenguin  : Committed with new changes....

File: avocado/utils/memory.py (hunk 2 above)
Comment (PraveenPenguin) [inline avocado/utils/memory.py]: i think avocado using pylint

Comment (apahim) [inline avocado/utils/memory.py]: We don't limit the line length to 80 characters in Avocado, but it is a good practice. If it does not compromise the readability, the change is welcomed.

File: avocado/utils/genio.py [hunk 4]
```diff
@@ -174,3 +174,16 @@ def write_one_line(filename, line):
     :type line: str
     """
     write_file(filename, line.rstrip('\n') + '\n')
+
+
+class GenIOException(Exception):
```
Comment (ldoktor) [inline avocado/utils/genio.py]: Would you please move this exception class on top of the module and add docstring about it's usage? It significantly improves readability.

File: avocado/utils/genio.py [hunk 5]
```diff
@@ -174,3 +174,16 @@ def write_one_line(filename, line):
...
+    pass
+
+
+def write_file_or_log(filename, data):
+    fd = os.open(filename, os.O_WRONLY)
+    try:
+        ret = os.write(fd, data)
+    except OSError as details:
+        raise GenIOException("The write to %s failed: %s" % (
```
Comment (ldoktor) [inline avocado/utils/genio.py]: nice

File: avocado/utils/memory.py (hunk 2 above)
Comment (ldoktor) [inline avocado/utils/memory.py]: Ideally this should be in a different patch, though (I can recommend `git cola` or `git citool` for that ;-) )

Comment (ldoktor) [inline avocado/utils/memory.py]: ^^ different patch, but could be in the same PR...

File: avocado/utils/memory.py [hunk 6]
```diff
@@ -278,3 +280,33 @@ def get_buddy_info(chunk_sizes, nodes="all", zones="all"):
...
+    :Param feature : Thp feature to get value
+    :type feature : str
+    """
+    thp_path = '/sys/kernel/mm/transparent_hugepage/'
+    thp_feature_to_get = os.path.join(thp_path, feature)
+    value = genio.read_file(thp_feature_to_get)
+    if feature in ("enabled", "defrag", "shmem_enabled"):
+        return (re.search(r"\[(\w+)\]", value)).group(1)
+    else:
+        return value
```
Comment (ldoktor) [inline avocado/utils/memory.py]: Nice and well structured.

File: avocado/utils/genio.py [hunk 7]
```diff
@@ -174,3 +174,16 @@ def write_one_line(filename, line):
     :type line: str
     """
     write_file(filename, line.rstrip('\n') + '\n')
+
+
+class GenIOException(Exception):
+    pass
+
+
+def write_file_or_log(filename, data):
```
Comment (ldoktor) [inline avocado/utils/genio.py]: Oh and would you please also describe what this does? (I mean the difference between this and the other classes)

Comment (ldoktor) [inline avocado/utils/genio.py]: Actually in this version it's function is not `write_file_or_log` but `write_file_or_fail` as you changed it to `raise Exception`.

Comment (ldoktor) [inline avocado/utils/genio.py]: ^^ maybe `write_file_or_crash`, not sure which is better.

Comment (ldoktor) [inline avocado/utils/genio.py]: Also in the docstring you could add `:raises GenIOException: On write failure`.

**Comment by santhosh93** [comment]
Opened a new PR @ https://github.com/avocado-framework/avocado/pull/1808
